runtime/common/src/migrations/nuke.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+#[cfg(feature = "try-runtime")]
+use frame_support::ensure;
+use frame_support::{
+	dispatch::GetStorageVersion,
+	storage::unhashed,
+	traits::{Get, OnRuntimeUpgrade, PalletInfoAccess, StorageVersion},
+	weights::{RuntimeDbWeight, Weight},
+};
+#[cfg(feature = "try-runtime")]
+use sp_std::vec::Vec;
+
+/// This upgrade nukes all storages from the pallet individually.
+/// This upgrade is only executed if pallet version has changed.
+///
+/// To handle possible issues forgeting removing the upgrade,
+/// you must specify the ON_CHAIN_VERSION,
+/// which represent the expected previous on-chain version when the upgrade is
+/// done. If these numbers mistmatch, the upgrade will not take effect.
+pub struct Migration<Pallet, DbWeight, const ON_CHAIN_VERSION: u16>(
+	sp_std::marker::PhantomData<(Pallet, DbWeight)>,
+);
+
+impl<Pallet, DbWeight, const ON_CHAIN_VERSION: u16> OnRuntimeUpgrade
+	for Migration<Pallet, DbWeight, ON_CHAIN_VERSION>
+where
+	Pallet: GetStorageVersion + PalletInfoAccess,
+	DbWeight: Get<RuntimeDbWeight>,
+{
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+		ensure!(
+			Pallet::on_chain_storage_version() == StorageVersion::new(ON_CHAIN_VERSION),
+			"Pallet on-chain version must match with ON_CHAIN_VERSION"
+		);
+
+		ensure!(
+			Pallet::on_chain_storage_version() < Pallet::current_storage_version(),
+			"Pallet is already updated"
+		);
+
+		ensure!(
+			unhashed::contains_prefixed_key(&pallet_prefix::<Pallet>()),
+			"Pallet prefix doesn't exists"
+		);
+
+		Ok(Vec::new())
+	}
+
+	fn on_runtime_upgrade() -> Weight {
+		if Pallet::on_chain_storage_version() != StorageVersion::new(ON_CHAIN_VERSION) {
+			log::error!(
+				"Nuke-{}: nuke aborted. This upgrade must be removed!",
+				Pallet::name()
+			);
+			return Weight::zero();
+		}
+
+		if Pallet::on_chain_storage_version() < Pallet::current_storage_version() {
+			log::info!("Nuke-{}: nuking pallet...", Pallet::name());
+
+			// TODO: How we can set the maximum for this? Currently hardcode
```

Comment:
I would set no Maximum TBH. If you want to nuke it, you need to ierate everything. WDYT?

Reply:
How can be protected against that, or even know that limit well taking into account that this will be done in the same block after the upgrade with other migrations?

Reply:
Migrations anyways block a whole block...

If it is too big, it is too big. That is currently the sad truth I would say.

Reply:
Mmm... so the following could not halt the chain:
- Upgrade runtime version.
  - Runtime is upgraded.
  - Each migration is done...
  - Huge nuke migration is done (suppose it would take 10 blocks of time)
- The node appears as "halted" so another node is chosen to create the runtime upgrade block, which later would appear halted too, etc...

Reply:
If the realy-chain can not produce the block in timely fashion or the parachain can`t we are stuck and must revert through relay democracy.

Reply:
> If [either] the [assigned] relay-chain [validators] or the parachain [collators] can not produce the block in timely fashion, we are stuck and must revert through relay democracy.

Made some minor additions. If the block cannot be produced, it will be tried over and over again as the upgrade is part of `on_initialize`. Therefore, if we either know the number of migrating items is sufficiently large or `try-runtime` returns a weight close enough to the maximum block weight, another approach needs to be chosen. 

Parity is [currently working on multi block migrations](https://github.com/paritytech/substrate/pull/14275). Until it is merged and tested and our chain updated to the supported substrate version, we should go with a lazy migration instead, if necessary. I provided some food for thoughts [here](https://github.com/centrifuge/centrifuge-chain/pull/1343#issuecomment-1591333784).

Reply:
Removed the maximum limit then 👍🏻